meteor/iterative.py

```diff
@@ -0,0 +1,270 @@
+    convergence_tolerance: float = 1e-4,
+    max_iterations: int = 1000,
+) -> tuple[np.ndarray, pd.DataFrame]:
+    """
+    Estimate the derivative phases using the iterative TV algorithm.
+
+    This function contains the algorithm logic.
+
+    Parameters
+    ----------
+    complex_native: np.ndarray
+        The complex native structure factors, usually experimental amplitudes and calculated phases
+
+    initial_complex_derivative : np.ndarray
+        The complex derivative structure factors, usually with experimental amplitudes and esimated
+        phases (often calculated from the native structure)
+
+    tv_denoise_function: Callable[[np.ndarray], tuple[np.ndarray, TvDenoiseResult]]
+        A function capable of applying the TV denoising operation to *Fourier space* objects. This
+        function should therefore map one complex np.ndarray to a denoised complex np.ndarray and
+        the TvDenoiseResult for that TV run.
+
+    convergance_tolerance: float
+        If the change in the estimated derivative SFs drops below this value (phase, per-component)
+        then return
+
+    max_iterations: int
+        If this number of iterations is reached, stop early.
+
+    Returns
+    -------
+    estimated_complex_derivative: np.ndarray
+        The derivative SFs, with the same amplitudes but phases altered to minimize the TV.
+
+    metadata: pd.DataFrame
+        Information about the algorithm run as a function of iteration. For each step, includes:
+        the tv_weight used, the negentropy (after the TV step), and the average phase change in
+        degrees.
+    """
+
+    derivative = np.copy(initial_derivative)
+    difference = initial_derivative - native
+
+    converged: bool = False
+    num_iterations: int = 0
+    metadata: list[dict[str, float]] = []
+
+    while not converged:
+        difference_tvd, tv_metadata = tv_denoise_function(difference)
+        updated_derivative = _project_derivative_on_experimental_set(
+            native=native,
+            derivative_amplitudes=np.abs(derivative),
+            difference=difference_tvd,
+        )
+
+        phase_change = average_phase_diff_in_degrees(derivative, updated_derivative)
+        derivative = updated_derivative
+        difference = derivative - native
+
+        converged = phase_change < convergence_tolerance
+        num_iterations += 1
+
+        metadata.append(
+            {
+                "iteration": num_iterations,
+                "tv_weight": tv_metadata.optimal_lambda,
+                "negentropy": tv_metadata.optimal_negentropy,
+                "average_phase_change": phase_change,
+            }
+        )
+
+        if num_iterations > max_iterations:
+            break
+
+    return derivative, pd.DataFrame(metadata)
+
+
+def iterative_tv_phase_retrieval(
+    input_dataset: rs.DataSet,
+    *,
+    native_amplitude_column: str = "F",
+    derivative_amplitude_column: str = "Fh",
```

Comment:
this seems as good as it can be to me right now. What were your concerns with the input/output?